Customizer: Add widget blocks section.

epiqueras opened the pull request:
cc @aduth @jorgefilipecosta 

Closes #13205

## Description

This PR adds the widget blocks editor to a new section in the Customizer. The `wp-edit-widgets` package that powers the new widgets page in WPAdmin has been used for this.

This is still a work in progress as `wp-edit-widgets` appears to have compatibility issues with legacy widgets. Some combinations of legacy widgets and widget blocks in some themes, e.g. Twenty Nineteen, break the Customizer or sometimes don't save the widgets for the right widget content area. All of the issues seem to be related to how `wp-edit-widgets` persists data.

After those issues are resolved, more work needs to be done to make the live preview work without page refreshes.

### Updates

- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-502892698
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-503190867
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-503211430
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-505557909
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-505656864
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-505665565
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-505933478
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-505963455
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-506858279
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-507386060
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-507690822
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-507875331
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-508215549
- https://github.com/WordPress/gutenberg/pull/16204#issuecomment-508523814

## How has this been tested?

This has only been tested manually.

## Screenshots

![Customizer Widget Blocks Editor](https://user-images.githubusercontent.com/19157096/59634937-839dac00-9115-11e9-8c91-b3e4a10e6467.png)

## Types of Changes

*New Feature:* Add a widget blocks section to the Customizer.

## Checklist:

- [ ] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.
- [x] I've included developer documentation if appropriate.

epiqueras committed:
Customizer: Add widget blocks section.

epiqueras committed:
Customizer: Fix incompatibility with legacy widgets sanitization.

epiqueras commented:
The issues have been fixed. They were caused by an incompatibility between how widget blocks are stored and legacy widgets sanitization for the Customizer. Next up, live-preview enhancements!

Relevant commit:

>https://github.com/WordPress/gutenberg/pull/16204/commits/4899f9a73eef983e7d26f0b7a29a7a4688b02e24 Customizer: Fix incompatibility with legacy widgets sanitization.

gziolo commented on .gitignore:
```diff
@@ -12,6 +12,7 @@ phpcs.xml
 yarn.lock
 docker-compose.override.yml
 /wordpress
+.vscode
```
You can create a global gitignore file instead to so you don't have to include IDE specific ignores to each project:
https://help.github.com/en/articles/ignoring-files#create-a-global-gitignore

We would have to keep a long list of IDE specific files for each WordPress project instead :)

epiqueras replied:
Done, thanks 👍

gziolo commented on lib/class-wp-customize-widget-blocks-control.php:
```diff
@@ -0,0 +1,32 @@
+ * Start: Include for phase 2
+ *
+ * @package gutenberg
+ * @since x.x.x
```
Next planned Gutenberg release is 6.0.

epiqueras replied:
👍

youknowriad commented on lib/widgets-page.php:
```diff
@@ -58,15 +58,21 @@ function gutenberg_widgets_init( $hook ) {
 		sprintf(
-			'wp.editWidgets.initialize( "widgets-editor", %s );',
+			'window.addEventListener( "load", function() {
+				wp.editWidgets.initialize( "widgets-editor", %s );
```
I think the theory here is that `editWidgets` is specific to the widgets screen and we didn't initially think of using it in the customizer directly.

I can see how it's a quick way to arrive to something usable, the issue though is that it forces us to play with CSS to have a sidebar with the "save" button... 

I think ideally, we'd:

 - Extract the common parts between the customizer and the widgets screen into a `@wordpress/widgets` or `@wordpress/widget-areas` package 
 - Build a new `widgets-customizer-panel` package that use the common parts in order to bootstrap the customizer package.
 - Try to merges the save button with the customizer's save button. (We shouldn't need two separate saving actions)

epiqueras replied:
What if we do that all in the same, current package. Merging the save buttons when rendering in the customizer should be simpler than splitting this into multiple packages.

epiqueras replied:
cc @jorgefilipecosta

jorgefilipecosta commented on lib/widgets.php:
```diff
@@ -213,3 +213,23 @@ function gutenberg_create_wp_area_post_type() {
 add_action( 'init', 'gutenberg_create_wp_area_post_type' );
 
 add_filter( 'sidebars_widgets', 'Experimental_WP_Widget_Blocks_Manager::swap_out_sidebars_blocks_for_block_widgets' );
+
```
I guess we can move this logic to a customizer.php file. 
For reference, this code was added because of the global widgets global used in this function:
```
	public function sanitize_sidebar_widgets_js_instance( $widget_ids ) {
		global $wp_registered_widgets;
		$widget_ids = array_values( array_intersect( $widget_ids, array_keys( $wp_registered_widgets ) ) );
		return $widget_ids;
	}
```
Does not contains the old format but contains the new format, I guess the best solution is to undestand why the global does not contain the backcompatible format and fix that problem as it may cause problems with plugins. If that is not possible I guess we can use this filter but instead of unsetting sanitize_js_callback I think we can make it point to new function where we compute the correct inputs based on the post id and we pass this values to the function that was set in $args['sanitize_js_callback'] before.
This does not seem a problem form this PR so I think we can address it separately provided we have a "Todo:" note to not miss this task.

epiqueras replied:
Yeah I wrote that in the documentation:

https://github.com/WordPress/gutenberg/blob/4899f9a73eef983e7d26f0b7a29a7a4688b02e24/lib/widgets.php#L218-L221

But, the issue is that `$widget_ids` is an int when it should be an array.

This filter, https://github.com/WordPress/gutenberg/blob/4899f9a73eef983e7d26f0b7a29a7a4688b02e24/lib/class-experimental-wp-widget-blocks-manager.php#L341, fixes that, but it doesn't run for that part of core.

I've moved it to customizer.php and added a TODO.

jorgefilipecosta reviewed (commented):
Hi @epiqueras, thank you for working on this task! There is good progress here in a short amount of time :)
I guess having a separate package for the customizer is going to be beneficial down the road as @youknowriad suggested.
I think we should also have a customizer.php where we add the customizer specific PHP code.

epiqueras committed:
Customizer: Set `@since` versions for new PHP methods.

Also remove the `.vscode` entry in .gitignore.

epiqueras commented:
@gziolo 's feedback.

Relevant commit:

>https://github.com/WordPress/gutenberg/pull/16204/commits/65360e1f9afd2150936d7e51aa7fb775e8d5fcaa Customizer: Set `@since` versions for new PHP methods.

epiqueras commented on lib/widgets-page.php:
```diff
@@ -58,15 +58,21 @@ function gutenberg_widgets_init( $hook ) {
-		'wp.blocks.unstable__bootstrapServerSideBlockDefinitions(' . wp_json_encode( get_block_editor_server_block_settings() ) . ');'
-	);
+
+	if ( 'gutenberg_customizer' !== $hook ) { // `get_block_editor_server_block_settings` is undefined in the Customizer.
```
Adding the comment for now.

youknowriad replied:
What does this mean exactly? Block registered server side won't be available in the customizer?

youknowriad replied:
Can't we just "include" the file where it's defined?

epiqueras replied:
>What does this mean exactly? Block registered server side won't be available in the customizer?

Yes

>Can't we just "include" the file where it's defined?

Huh, I got rid of the if guard and it worked. I don't see any obvious changes in Core that would have fixed this 🤔

epiqueras replied:
https://github.com/WordPress/gutenberg/pull/16204#issuecomment-507690822

epiqueras commented on gutenberg.php:
```diff
@@ -68,6 +68,32 @@ function gutenberg_menu() {
+ * @param \WP_Customize_Manager $wp_customize An instance of the class that controls most of the Theme Customization API for WordPress 3.4 and newer.
+ * @since x.x.x
+ */
+function gutenberg_customize_register( $wp_customize ) {
```
Good idea.

epiqueras commented on lib/widgets-page.php:
```diff
@@ -58,15 +58,21 @@ function gutenberg_widgets_init( $hook ) {
 		'wp-edit-widgets',
 		sprintf(
-			'wp.editWidgets.initialize( "widgets-editor", %s );',
+			'window.addEventListener( "load", function() {
```
Awesome, didn't know we had this!

epiqueras committed:
Customizer: Refactor new PHP code so that it can be loaded conditionally.

Also add missing TODOs.

epiqueras commented:
@jorgefilipecosta 's feedback.

Relevant commit:

>https://github.com/WordPress/gutenberg/pull/16204/commits/7fa5a38d270295d818a3b431da5c8031d68d7f75 Customizer: Refactor new PHP code so that it can be loaded conditionally.
>
>Also add missing TODOs.

The next tasks would be not rendering the `Update` button when on the Customizer and integrating into the Customizer's save and live preview functionality.

epiqueras committed:
Customizer: Integrate widget blocks live preview and saving.

epiqueras commented:
Saving customizer changes now also saves widget block edits and live preview is working:

>https://github.com/WordPress/gutenberg/pull/16204/commits/271a49f850d12f3bb8db2ed48b8ac90a05d246b6 Customizer: Integrate widget blocks live preview and saving.

epiqueras commented:
![Kapture 2019-06-25 at 13 18 54](https://user-images.githubusercontent.com/19157096/60122994-06dd8400-974c-11e9-90d4-972ae772f377.gif)

youknowriad reviewed (commented):
I took this for a spin, and it's great. I feel with the saving and previewing we kind of addressed the most difficult challenges of this screen.

Some things that would be good to improve.

 - I think the "Inspector" should probably be shown in a modal instead of a sidebar in that screen. This is doable by adding a button to toggle the modal to the Block settings menu (there's a lot for that). By I feel it could be left for a separate PR and just remove the inspector from the initial implementation.

 - Instead of reusing `wp.editWidgets.initialize( "widgets-editor", %s );` and trying to detect whether we're the customizer or widgets screen using DOM or availability of globals, I think we should have:
     - a separate wp.editCustomizerWidgetsPanel function to render a separate React Tree that is directly optimized for the customizer screen.
     - reuse as much as we can from the edit-widgets Tree of components (WidgetArea...)
     - Avoid rendering the Inspector in that customizer tree (we could use a modal but could be a separate PR)

- I think there are some z-index issues for the block toolbar and styling issues for the areas (the width is too small by default) but these tweaks can also be addressed separately.

What do you think?

youknowriad commented:
I also had some PHP notices while saving the areas

```
Notice: Undefined index: blocks-widget-4ad2ac37d121d5a87a8176a54caa2e5e in /var/www/html/wp-includes/class-wp-customize-widgets.php on line 497 
Notice: Undefined index: blocks-widget-4ad2ac37d121d5a87a8176a54caa2e5e in /var/www/html/wp-includes/class-wp-customize-widgets.php on line 509 
Notice: Undefined index: blocks-widget-4ad2ac37d121d5a87a8176a54caa2e5e in /var/www/html/wp-includes/class-wp-customize-widgets.php on line 510 
Notice: Undefined index: blocks-widget-4ad2ac37d121d5a87a8176a54caa2e5e in /var/www/html/wp-includes/class-wp-customize-widgets.php on line 579 
Notice: Undefined index: blocks-widget-ccd122757280cfba6bd1804a475deccc in
```

epiqueras commented:
@youknowriad 

>I think the "Inspector" should probably be shown in a modal...

Yeah, that would look a lot better. I'll follow up this PR with that as suggested.

>Instead of reusing wp.editWidgets.initialize...

This makes a lot of sense, specially because the current tree is still being modified in parallel, but what if we have `wp.editWidgets.initializeWidgetsScreen` and `wp.editWidgets.initializeWidgetsCustomizerSection`. It seems like that would make it simpler to share the code.

>I think there are some z-index issues for the block toolbar...

I'll fix those in the new component tree in this PR, should be simple.

>I also had some PHP notices while saving the areas...

Looks like new Customizer PHP incompatibilities with post IDs being used as sidebar content. There's probably a hook to bailout from whatever it's trying to do, I'll look into it after the other changes.

epiqueras committed:
Customizer: Refactor widget blocks editor into a new component tree.

epiqueras committed:
Customizer: Allow block toolbars to overflow into preview.

epiqueras commented:
@youknowriad 's feedback

>https://github.com/WordPress/gutenberg/pull/16204/commits/0643ace2fd328a3d2eb769b4f1d2e24bcfbfef5f Customizer: Refactor widget blocks editor into a new component tree.

>https://github.com/WordPress/gutenberg/pull/16204/commits/40923806c5b8da87af63584a29eb916c9f83b066 Customizer: Allow block toolbars to overflow into preview.

epiqueras committed:
Customizer: Fix PHP notices due to missing widget controls.

epiqueras commented:
Wrapped up.

>https://github.com/WordPress/gutenberg/pull/16204/commits/e41149b79b4cf2b3be9d6ad80f066b49cdb2e77c Customizer: Fix PHP notices due to missing widget controls.

youknowriad commented:
Thanks @epiqueras, this is looking better. A few remarks:

 - It feels like the "scheduling" and "drafting" of customizer changes is broken because we're always publishing the changes no matter what button we click. I guess we need to find a way to hook into the "changeset" API of the customizer.

 - when the content of the widget area is long, I'm not able to scroll the sidebar (not really critical and can be solved separately I think)

 - I feel like snackbars are not needed here. this is probably a symptom of the way we're saving the changes now (intead of using the customizer changesets, we trigger a separate API request)

epiqueras committed:
Customizer: Remove notices and overflow hack. Block toolbar should wrap instead.

epiqueras commented:
@youknowriad 

>It feels like the "scheduling" and "drafting" of customizer changes...

Changes are only published when the publish button is clicked, everything else is just a preview.

>when the content of the widget...

This happened because I changed things to allow the block toolbar to overflow out of the panel horizontally. Browser specs disallow horizontal overflows with vertical scrolling. I've removed the bit of CSS that did this. I think the correct solution is for the toolbar to wrap properly.

>I feel like snackbars are not needed here...

Yeah, it looks better now without them.

>https://github.com/WordPress/gutenberg/pull/16204/commits/d2c4f53d6507f3478e96d6bed8339a9240499b88 Customizer: Remove notices and overflow hack. Block toolbar should wrap instead.

epiqueras commented on lib/class-experimental-wp-widget-blocks-manager.php:
```diff
@@ -330,6 +330,12 @@ public static function convert_blocks_to_widget( $blocks ) {
 		);
+		wp_register_widget_control(
+			$widget_id,
+			__( 'Blocks Area ', 'gutenberg' ),
```
Yeah, I just copied it from the other usage, fixing both now.

epiqueras commented on lib/class-wp-customize-widget-blocks-control.php:
```diff
@@ -0,0 +1,32 @@
+<?php
+/**
+ * Start: Include for phase 2
```
I agree. Changing it.

epiqueras commented on lib/class-wp-customize-widget-blocks-control.php:
```diff
@@ -0,0 +1,32 @@
+ * Start: Include for phase 2
+ *
+ * @package gutenberg
+ * @since 6.0.0
```
I'll update it once we are ready to merge.

epiqueras commented on lib/widgets-page.php:
```diff
@@ -9,10 +9,21 @@
  *
  * @since 5.2.0
+ *
+ * @param string $page Page.
```
Updating comment to:

>The page name the function is being called for, `'gutenberg_customizer'` for the Customizer.

epiqueras commented on lib/widgets-page.php:
```diff
@@ -9,10 +9,21 @@
+ * @param string $page Page.
  */
-function the_gutenberg_widgets() {
+function the_gutenberg_widgets( $page ) {
```
You are right, it just registers it to an action with the name. I've made it default to the widgets screen for future proofing it for edits that expect it.

epiqueras commented on packages/edit-widgets/src/utils/with-wp-customize.js:
```diff
@@ -0,0 +1,36 @@
+ */
+import { serialize } from '@wordpress/blocks';
+
+/* global wp */
```
That makes sense, don't know why I didn't think of it.

epiqueras commented on packages/edit-widgets/src/utils/with-wp-customize.js:
```diff
@@ -0,0 +1,36 @@
+
+/* global wp */
+
+function previewBlocksInWidgetArea( id, blocks ) {
```
Done.

epiqueras commented on packages/edit-widgets/src/utils/with-wp-customize.js:
```diff
@@ -0,0 +1,36 @@
+		'#customize-preview > iframe'
+	);
+
+	if ( customizePreviewIframe && customizePreviewIframe.contentDocument ) {
```
I usually switch between both approaches, but I think my new rule will be to never nest if statements.

epiqueras commented on packages/edit-widgets/src/utils/with-wp-customize.js:
```diff
@@ -0,0 +1,36 @@
+		);
+
+		if ( widgetArea ) {
+			widgetArea.innerHTML = serialize( blocks );
```
I doubt it will be a problem with the kind of blocks that go here, but I added a 1 second throttle just in case.

epiqueras commented on packages/edit-widgets/src/utils/with-wp-customize.js:
```diff
@@ -0,0 +1,36 @@
+
+	if ( customizePreviewIframe && customizePreviewIframe.contentDocument ) {
+		const widgetArea = customizePreviewIframe.contentDocument.querySelector(
+			`[data-customize-partial-placement-context*='"sidebar_id":"${ id }"']`
```
Yeah, otherwise you need to escape the inner ones.

epiqueras commented on packages/edit-widgets/src/index.js:
```diff
@@ -27,3 +28,19 @@ export function initialize( id, settings ) {
+ * @param {string} id       ID of the root element to render the section in.
+ * @param {Object} settings Block editor settings.
+ */
+export function customizerInitialize( id, settings ) {
```
>This would assume that widgets behavior is pretty common across the customizer and widgets screens.

It is.

>One potential downside to this is that we load this entire module for both the widgets and customizer screens, even if large parts of code which is loaded isn't actually relevant for the current screen.

Only 2 small files are not relevant, everything else is the same. Do you think there is something that could be pruned there?

epiqueras commented on packages/edit-widgets/src/components/customizer-layout/style.scss:
```diff
@@ -0,0 +1,22 @@
+	padding: 30px 0;
+
+	.block-editor {
+		&-block-list {
```
I find this easier to read. But I didn't think that it could break search. I think that alone is enough reason to not nest it.

Editors should have a setting for parsing nested selectors in their search...

epiqueras commented on packages/edit-widgets/src/components/customizer-layout/index.js:
```diff
@@ -0,0 +1,44 @@
+
+function CustomizerLayout( { blockEditorSettings } ) {
+	return (
+		<SlotFillProvider>
```
Maybe a general `EditorProvider`?

epiqueras committed:
Customizer: Optimize and clean up changes.

- Remove extra whitespace in widget names.
- Add JSDocs and improve comments.
- Add missing default parameter.
- Flatten CSS selectors.
- Throttle live preview updates.

epiqueras commented:
@aduth 's feedback

>https://github.com/WordPress/gutenberg/pull/16204/commits/870d6af49db9cf930446c1e7ff26ccf9a5a452e9 Customizer: Optimize and clean up changes.
>
>- Remove extra whitespace in widget names.
>- Add JSDocs and improve comments.
>- Add missing default parameter.
>- Flatten CSS selectors.
>- Throttle live preview updates.

jorgefilipecosta commented on packages/edit-widgets/src/utils/with-wp-customize.js:
```diff
@@ -0,0 +1,36 @@
+		);
+
+		if ( widgetArea ) {
+			widgetArea.innerHTML = serialize( blocks );
```
I think this a great simple first step that allows us to try the block preview and is useful for testing with users.
Unfortunately, this method does not work with dynamic blocks. I guess we can use the ServerSideRender component to have a preview for these blocks.
Maybe we can serialize these blocks differently, e.g., a container div, and then we call React render on that div to render a ServerSideRender component?
I guess if we get this method working with dynamic blocks, we will be able to have widget previews as they are a dynamic block.

epiqueras replied:
That sounds like the way to go. I looked into the Customizer PHP API for this, but it is very opinionated on the format of the data that gets sent to the previewer. It looks like it expects widget form data.

epiqueras replied:
We could have an async serializer that uses `ServerSideRender` to render dynamic blocks. Should I go ahead with that?

@aduth I think you know a lot about the serializer right?

jorgefilipecosta replied:
I guess we can leave the preview of dynamic blocks for the second step in a follow-up. The objective of this phase is to have something users can test, and I think dynamic blocks are not a blocker. What do you think @youknowriad?

If we follow the logic of an async serializer it would be even more important that we don't serialize a block again if no changes happened. For each server-side rendered block we may have database queries and we should avoid making a server request each second. Maybe the problem is as simple as applying memoization on the serialization? (we would memoize based on block uid and the attributes object, so if the attributes did not change we would use the previous serialization result without any new request /rerender, not used memoization keys during serialization would be removed to avoid memory leaks or maybe a weakmap can also help us here).

epiqueras replied:
Yeah, memoization would do the trick.

jorgefilipecosta commented on lib/class-experimental-wp-widget-blocks-manager.php:
```diff
@@ -330,6 +330,12 @@ public static function convert_blocks_to_widget( $blocks ) {
 				'blocks' => $blocks,
 			)
 		);
+		wp_register_widget_control(
```
According to the commit, it seems like this was added to avoid missing widget control errors. The control is the edition interface of a widget, and it looks like widget controls are not mandatory, e.g., I can have a print hello widget, without any UI to edit it.
It would be interesting to check if the warnings also happen on other widgets without a control associated as it may be an existing core bug or it may be another problem in our Blocks Area Widget.
For now, I guess if this solution fixes the problem it is something we can use.

epiqueras replied:
Yeah I found it strange too.

jorgefilipecosta commented on packages/edit-widgets/src/components/customizer-layout/index.js:
```diff
@@ -0,0 +1,44 @@
+					blockEditorSettings={ blockEditorSettings }
+					onUpdateBlocks={ useCallback(
+						( id, blocks ) =>
+							withWPCustomize( ( { previewBlocksInWidgetArea, saveButton } ) => {
```
I guess customizer was added because for customizer we need to call previewBlocksInWidgetArea. What if instead on our customizer code we subscribe to the widgets store `wp.data.subscribe('core/edit-widgets')` and then when an area changes we update its preview? I guess it would also provide more isolation as we would not have the widgets module calling/interacting with the dom/external code, instead external code would react to changes in the widget store.
Do you think it would simplify things or do you predict some disadvantage?

epiqueras replied:
It's also needed for styling right now. We could also just have the extra class toggled through a prop, but I wouldn't be surprised if we do more layout changes on the widgets screen that shouldn't also affect the Customizer section.

That subscription approach would work, although you would have to subscribe to the whole registry (it only supports global listeners), run the relevant selectors, and call a memoized function with the blocks. All in all, that makes the code way more complex. I don't think unavoidable DOM manipulations like these in a React tree are bad as long as they are minimized and only run on side effects or interactions. But, if we have a pre existing convention in the rest of the codebase, I can change it.

epiqueras replied:
I ended up going with the manual subscription approach, because the complexity that came with adding scheduling, drafting, and share URLs warranted it.

youknowriad commented:
> It feels like the "scheduling" and "drafting" of customizer changes...

I was talking about this button, where you can change the purpose of the publish button.

<img width="602" alt="Capture d’écran 2019-06-27 à 8 16 41 AM" src="https://user-images.githubusercontent.com/272444/60245208-03fda300-98b4-11e9-9433-fa6003b64421.png">

If you make an edit to something else in the customizer (say site title), it will show up. But for some reason, it's not showing when you edit the block areas widgets directly. But you can still edit the site title, then edit the block widgets and you could click "save draft" or "schedule" but right now they don't work as expected.

epiqueras commented:
> > It feels like the "scheduling" and "drafting" of customizer changes...
> 
> I was talking about this button, where you can change the purpose of the publish button.
> 
> <img alt="Capture d’écran 2019-06-27 à 8 16 41 AM" width="602" src="https://user-images.githubusercontent.com/272444/60245208-03fda300-98b4-11e9-9433-fa6003b64421.png">
> 
> If you make an edit to something else in the customizer (say site title), it will show up. But for some reason, it's not showing when you edit the block areas widgets directly. But you can still edit the site title, then edit the block widgets and you could click "save draft" or "schedule" but right now they don't work as expected.

Ohh, I missed that, why would it not show for me?

https://user-images.githubusercontent.com/19157096/59634937-839dac00-9115-11e9-8c91-b3e4a10e6467.png

epiqueras commented:
>I was talking about this button, where you can change the purpose of the publish button.

Ohh, I missed that. I think that only shows when making other Customizer changes in parallel. Changeset drafting and scheduling is tightly coupled in the Customizer API, should I somehow work this into that API, or should we work on new ones?

youknowriad commented:
> should I somehow work this into that API

If we can, I think so yes.

epiqueras committed:
Customizer: Sync Changesets

Widget area edits made in the Customizer are synced to Customizer
changesets as an object, encoded as a JSON string, where the keys
are widget area IDs and the values are serialized block content.
This file takes care of that syncing using the 2-way data binding
supported by `WP_Customize_Control`s. The process is as follows:

- On load, the client checks if the current changeset has
widget areas that it can parse and use to hydrate the store.
It will load all widget areas for the current theme, but if
the changeset has content for a given area, it will replace
its actual published content with the changeset's.

- On edit, the client updates the 2-way bound input with a new object that maps
widget area IDs and the values are serialized block content, encoded
as a JSON string.

- On publish, a PHP action will parse the JSON string in the
changeset and update all the widget areas in it, to store the
new content.

epiqueras commented:
>https://github.com/WordPress/gutenberg/pull/16204/commits/5ad559fd6f05db99f604903352fefcc827bcd951 Customizer: Sync Changesets
>
>Widget area edits made in the Customizer are synced to Customizer
>changesets as an object, encoded as a JSON string, where the keys
>are widget area IDs and the values are serialized block content.
>This file takes care of that syncing using the 2-way data binding
>supported by `WP_Customize_Control`s. The process is as follows:
>
>- On load, the client checks if the current changeset has
>widget areas that it can parse and use to hydrate the store.
>It will load all widget areas for the current theme, but if
>the changeset has content for a given area, it will replace
>its actual published content with the changeset's.
>
>- On edit, the client updates the 2-way bound input with a new object that maps
>widget area IDs and the values are serialized block content, encoded
>as a JSON string.
>
>- On publish, a PHP action will parse the JSON string in the
>changeset and update all the widget areas in it, to store the
>new content.

youknowriad commented:
Something is not working properly with the saving.

 - Try editing a widget area
 - Save the changes using the "Save Draft" option.
 - Check the preview link you get.
 - Notice that the updates are not there.

epiqueras committed:
Customizer: Make widget blocks compatible with Changeset share URLs.

epiqueras commented:
@youknowriad I didn't realize that share URLs used different mechanisms. This commit makes widget blocks compatible:

>https://github.com/WordPress/gutenberg/pull/16204/commits/4308984949371bcbabfa68511ad5088e637e5671 Customizer: Make widget blocks compatible with Changeset share URLs.

youknowriad commented on packages/edit-widgets/src/components/customizer-edit-widgets-initializer/sync-customizer.js:
```diff
@@ -0,0 +1,120 @@
+			// Subscribe to registry changes so that we can update the preview
+			// and the hidden input's value when any of the widget areas change.
+			widgetAreas = getWidgetAreasObject();
+			window.wp.data.subscribe( () => {
```
Are we considered in terms of performance here? I feel we don't really need to support the live-preview, especially because the rendered HTML can be a lot different from the HTML used in the blocks. We can have dynamic blocks, we can have server-side hooks... I feel it's just going to be confusing. 

That said, I don't think it's a blocker for this PR.

epiqueras replied:
>Are we considered in terms of performance here?...

I think that with the throttling in place, the performance hit is minimal, although I haven't done any serious testing besides 6x CPU slowdown on my machine working fine.

>...especially because the rendered HTML can be a lot different from the HTML used in the blocks...

I have plans for another PR that adds an asynchronous `serialize` that uses `<ServerSideRender />` to render dynamic blocks. I could also look at time-slicing or offloading serialization to another thread and that would also take care of any performance concerns here.

youknowriad commented on packages/edit-widgets/src/store/actions.js:
```diff
@@ -15,9 +15,11 @@ const WIDGET_AREAS_SAVE_NOTICE_ID = 'WIDGET_AREAS_SAVE_NOTICE_ID';
  * @yields {Object} Action object.
  */
-export function* setupWidgetAreas() {
+export function* setupWidgetAreas( savedWidgetAreas ) {
```
I'm not sure why do we need this change here. Can't we just update the blocks in the store once we edit them? and rely on the fact that the store is always up to date?

epiqueras replied:
It's always up to date with the "published" widgets.

This allows us to override specific widget areas with their Changeset counterparts.

youknowriad replied:
I'm not certain I grasp the details yet but I think the `edit-widgets` store is used to track the current state of the editor, so maybe we should track the changeset couterpart instead of the published widgets. cc @jorgefilipecosta 

What value do we have in tracking the published widgets there?

epiqueras replied:
>...I think the edit-widgets store is used to track the current state of the editor...

Yes

`setupWidgetAreas` loads the published widget areas into the store when the app starts. What I did here was provide a way for the Customizer sync code to replace any combination of those widget areas with what is saved in the Changeset. The published widget areas are mapped into the objects used by the store, but their `blocks` property is set to the Changeset value, if defined:

https://github.com/WordPress/gutenberg/blob/4308984949371bcbabfa68511ad5088e637e5671/packages/edit-widgets/src/store/actions.js#L35

jorgefilipecosta replied:
It seems we are calling setupWidgetAreas just as a way to replace the blocks with the customizer changeset version. Can't we just do a loop call to updateBlocksInWidgetArea to replace blocks with customizer changeset version, avoiding the need to change this action?

epiqueras replied:
See https://github.com/WordPress/gutenberg/pull/16204#discussion_r299709914.

youknowriad commented on packages/edit-widgets/src/components/customizer-edit-widgets-initializer/sync-customizer.js:
```diff
@@ -0,0 +1,120 @@
+if ( window.wp && window.wp.customize && window.wp.data ) {
+	// Wait for the Customizer to finish bootstrapping.
+	window.wp.customize.bind( 'ready', () =>
+		window.wp.customize.previewer.bind( 'ready', () => {
```
I do wonder if we can attach these behaviors to an effect in the `Initializer` component for instance, I think it makes it a bit easier to reason about to think about.

epiqueras replied:
That was the original approach, but we pivoted a little bit here:

https://github.com/WordPress/gutenberg/pull/16204#discussion_r297832241

youknowriad reviewed (commented):
This seems almost ready for a v1.

I'd appreciate a second review @jorgefilipecosta @aduth and others.

epiqueras committed:
Customizer: Add server registered blocks back in. The getter function is now defined?

epiqueras commented:
>https://github.com/WordPress/gutenberg/pull/16204/commits/b3211fb8a89c7813469857fffc23f51eaed5f403 Customizer: Add server registered blocks back in. The getter function is now defined?

jorgefilipecosta commented on packages/edit-widgets/src/components/customizer-edit-widgets-initializer/sync-customizer.js:
```diff
@@ -0,0 +1,120 @@
+		`[data-customize-partial-placement-context*='"sidebar_id":"${ id }"']`
+	);
+	if ( widgetArea ) {
+		widgetArea.innerHTML = serialize( blocks );
```
Related to another comment, maybe if all the blocks are legacy widget blocks, we should not update the preview. Users that just have normal widgets and never used our new block mechanism may open the Block Widgets section, and when they do it, they will lose the widget previews right away even if they go back to the standard widget sections,  this check will reduce a little bit that problem.

epiqueras replied:
https://github.com/WordPress/gutenberg/pull/16204#discussion_r299705910

That fixes this too, but widget block changes will still take precedence over legacy widget ones. I think we should just show one or the other, having both is confusing.

jorgefilipecosta commented on lib/customizer.php:
```diff
@@ -0,0 +1,120 @@
+	);
+	$wp_customize->add_section(
+		'gutenberg_widget_blocks',
+		array( 'title' => __( 'Widget Blocks', 'gutenberg' ) )
```
Would it make sense to name this section 'Widget Blocks (Experimental)', to make users more aware of the experimental nature of this section? Similar to what we did with the widget blocks screen and the legacy widgets block.

epiqueras replied:
Yes

jorgefilipecosta reviewed (commented):
Hi @epiqueras, excellent work here, it is fantastic that you managed to find solutions that allow us to test complex feature like the customizer previews 👍
I think this is almost ready to merge; I left some comments that I think we should consider before the merge.
They are mostly changes to reduce the impact on users that are not trying the block widgets experiments.

jorgefilipecosta commented on lib/class-experimental-wp-widget-blocks-manager.php:
```diff
@@ -340,20 +346,34 @@ public static function convert_blocks_to_widget( $blocks ) {
 			self::$unfiltered_sidebar_widgets = $sidebars_widgets;
 		}
+		$changeset_data = null;
+		if ( is_customize_preview() ) {
```
Maybe a function exists check on is_customize_preview provides extra security to not break special sites that may not have this function defined.

epiqueras replied:
>It seems we are calling setupWidgetAreas just as a way to replace the blocks with the customizer changeset version. Can't we just do a loop call to updateBlocksInWidgetArea to replace blocks with customizer changeset version, avoiding the need to change this action?

Good catch! This is actually not needed anymore after the changes for supporting share URLs. Changeset data is already rendered on the server.

>Maybe a function exists check on is_customize_preview provides extra security to not break special sites that may not have this function defined.

Yes!

epiqueras replied:
>Changeset data is already rendered on the server.

Actually, no, because the store loads from an API that doesn't go through that filter, I'll go with the `updateBlocksInWidgetArea` approach after setup and before the subscription.

epiqueras commented on lib/class-experimental-wp-widget-blocks-manager.php:
```diff
@@ -330,6 +330,12 @@ public static function convert_blocks_to_widget( $blocks ) {
+		wp_register_widget_control(
+			$widget_id,
+			__( 'Blocks Area', 'gutenberg' ),
+			'noop',
```
I wonder why I didn't get that error.

I'll change it to 'echo'.

epiqueras commented on packages/edit-widgets/src/components/customizer-edit-widgets-initializer/sync-customizer.js:
```diff
@@ -0,0 +1,120 @@
+// Check that all the necessary globals are present.
+if ( window.wp && window.wp.customize && window.wp.data ) {
+	// Wait for the Customizer to finish bootstrapping.
+	window.wp.customize.bind( 'ready', () =>
```
I am changing it to run after the editor finishes setup so that it only fires after the first edit.

epiqueras committed:
Customizer: Fix legacy compatibility issues.

- Guard for undefined functions.
- Add "experimental" warning to section title.
- Don't let customizer syncing overwrite legacy widget previews until after the first edit.

epiqueras commented:
@jorgefilipecosta 's feedback

>https://github.com/WordPress/gutenberg/pull/16204/commits/a86e801f4d0c1bbd1a3bc5140e20090c643bcf8a Customizer: Fix legacy compatibility issues.
>
>- Guard for undefined functions.
>- Add "experimental" warning to section title.
>- Don't let customizer syncing overwrite legacy widget previews until after the first edit.

jorgefilipecosta commented on lib/class-experimental-wp-widget-blocks-manager.php:
```diff
@@ -330,6 +330,12 @@ public static function convert_blocks_to_widget( $blocks ) {
+		wp_register_widget_control(
+			$widget_id,
+			__( 'Blocks Area', 'gutenberg' ),
+			'echo',
```
This still triggers an error:
```
call_user_func_array() expects parameter 1 to be a valid callback, function 'echo' not found or invalid function name
````
It seems echo is not considered a callback function, but using a function that prints something would be dangerous anyway as printing something during REST request may make the requests invalid.
It seems our best bet is to create a simple noop function ourselves and pass its name here
.

epiqueras replied:
Will do that 👍

jorgefilipecosta reviewed (commented):
I'm noticing a bug that was not happening in the tests yesterday, not sure if it is something in my setup, it would be good to have a double check here:
If we use the normal widget screen to set up widgets and then we go to the customizer we delete the legacy widget blocks, we add some blocks (paragraphs) and we press publish right away without saving a draft first, the saving does not happen.

epiqueras committed:
Widget Blocks Manager: Define own noop function.

epiqueras commented:
>I'm noticing a bug that was not happening in the tests yesterday, not sure if it is something in my setup, it would be good to have a double check here:
If we use the normal widget screen to set up widgets and then we go to the customizer we delete the legacy widget blocks, we add some blocks (paragraphs) and we press publish right away without saving a draft first, the saving does not happen.

It works as expected for me. But I think one changeset has to be taking precedence over the other. I don't think showing both edit options at the same time is a good idea. It will confuse users at best, and cause bugs at worst. Can we hide the legacy option if the new one is enabled?

>https://github.com/WordPress/gutenberg/pull/16204/commits/87f2c377da413dc1429fcd2700f27723ddf3bd37 Widget Blocks Manager: Define own noop function.

jorgefilipecosta reviewed (approved):
Thank you for all the work, the iterations and diving deep on this feature 🥇. The initial objective was to have something experimental that we can test and iterate on, and I think we achieved this objective. This is a complex problem, and we have a good first step here.

I think we can merge (right after Travis is happy) and continue the development in other PR's 💥🎉.


Travis is identifying a lint problem on FILE: /app/lib/customizer.php. We should also update the "since versions" on some functions. Feel free to apply these changes and merge after.

For future reference/tests, I'm also writing here potential issues found during my tests. I don't think they are blockers.

Some issues I noticed that I think it is because we did not have yet a mechanism to handle the case:
- When someone does some changes in the widgets tab (e.g., add a new widget) even if they are saved as drafts our block widget section does not use it.
- When we preview another theme in the customizer, the standard widgets section updates to reflect the areas of the new theme but the new section doesn’t.

Things that may not be related/may be something on my setup/ plugins I have, etc:
- There is a strange problem if we add blocks in a widget section and then we reload the old widget screen two times the blocks get removed, and some widgets get used (may not be related to this PR, previously a similar problem was happening but just when a changed was applied on the widget screen).
- Adding blocks in the customizer on the 2014 theme may not work well. Problem described in more detail above; it works well when the area contains blocks, the problem only happens when the area contains standard widgets, and we are saving blocks for the first time (the problem seems intermittent).

Another issue I noticed is that if we add blocks to an area. We save everything; we reload the customizer we go to the standard widgets section; we delete the "Blocks Area" widget, and we try to re-add other widgets, the widget add does not work, and it is impossible for the user to re-add widgets using just the customizer. So if a user tries our experimental section and then wants to revert the only option is using the old widget screen, it is impossible using just the customizer, and some users may not be aware the widgets screen exist and use the customizer all the time. I guess this issue is the one with the most priority.

epiqueras committed:
Customizer: Update `@since` version numbers.

epiqueras commented:
>https://github.com/WordPress/gutenberg/pull/16204/commits/f1985b6350ee09c64573ab2370018a9f19eb54f0 Customizer: Update `@since` version numbers.

youknowriad commented:
Congrats for your first PR @epiqueras 🎉
